bazzani/kafka-connect-stack
# Add JDBCSourceConnector and JSON Schema validation

Ingesting data from a Database to a Kafka Topic is a common requirement
and the `JDBCSourceConnector` can do it for us by providing a number of
query modes (Incrementing/Timestamp/Custom Query):
 - https://docs.confluent.io/kafka-connectors/jdbc/current/source-connector/overview.html

We will use `Incrementing Column` mode for this simple example, but
others are available to provide more robust and accurate data ingestion.

In this change we add a MySQL database table as the source of the data,
and the `incrementing` mode of the connector to find new rows by id.

The `mysql-connector-j-8.4.0.jar` needed to be downloaded using wget to
the Connect docker image at build time:
- https://dev.mysql.com/downloads/connector/j/

This change also uses JSON schemas to ensure that only data compliant
against the JSON schema is produced to the Kafka Topic. Some connector
properties to support that are:

```json
"value.converter": "io.confluent.connect.json.JsonSchemaConverter",
"value.converter.schema.registry.url": "http://schema-registry:8081",
"value.converter.schemas.enable": "false",
"value.converter.auto.register.schemas": "false",
"value.converter.json.fail.invalid.schema": "true",
"value.converter.latest.compatibility.strict": "false",
"value.converter.use.latest.version": "true",
```

The JSON Schema restricts the name to 40 chars, and the age to 125. If
the JSON Schema validation fails, we want the connector to log the
validation error(s) (or any other errors) and continue to process data.

The connector and task do not fail or stop, unlike the default behaviour
a Sink connector would normally have. The following properties provide
this behaviour:

```json
"errors.tolerance": "all",
"errors.log.enable": "true",
"errors.log.include.messages": "true"
```

- see more at https://docs.confluent.io/platform/current/installation/configuration/connect/source-connect-configs.html#errors-tolerance

---

A @SpringBootTest has been added to test the connector:

1. add a row to the database with data valid against the JSON schema
2. add a row to the database with data invalid against the JSON schema

- The tests check for the presence/absence of the record on the topic.

Application properties have been added for the Kafka Consumer that will
consume records from the topic with a String key and a JSON value. The
`KafkaJsonSchemaDeserializer` is used to create a JSON property map
which is converted to a `UserData` POJO using the jackson ObjectMapper.

The `KafkaUserDataConsumerFixture` encapsulates the `@KafkaListener`
Spring logic and exposes the latest record consumed from the topic.

The test topic is created via `KafkaAdmin`/`NewTopic` beans at startup.
- see more at https://docs.spring.io/spring-kafka/reference/kafka/configuring-topics.html

The `SchemaRegistryClient` is used to register the JSON schema subject
for the test topic; this can also be achieved by making a cURL request:
- https://docs.confluent.io/platform/current/schema-registry/develop/api.html#post--subjects-(string-%20subject)-versions

---

A new Datasource has been added for mysql to the config file in `.idea`.

---

For more information about JSONSchemaConverter, converters in general
and how to use a JSON schema with a topic:
- https://docs.confluent.io/platform/7.6/schema-registry/fundamentals/serdes-develop/serdes-json.html#json-schema-compatibility-rules
- https://www.confluent.io/blog/kafka-connect-deep-dive-converters-serialization-explained/
- https://gist.github.com/rmoff/f32543f78d821b25502f6db49eee9259
- https://docs.confluent.io/platform/7.6/schema-registry/connect.html#json-schema
